agent/grpc/handler.go

```diff
@@ -0,0 +1,109 @@
+/*
+Package grpc provides a Handler and client for agent gRPC connections.
+*/
+package grpc
+
+import (
+	"fmt"
+	"net"
+
+	"google.golang.org/grpc"
+)
+
+// NewHandler returns a gRPC server that accepts connections from Handle(conn).
+func NewHandler(addr net.Addr) *Handler {
+	// We don't need to pass tls.Config to the server since it's multiplexed
+	// behind the RPC listener, which already has TLS configured.
+	srv := grpc.NewServer(
+		grpc.StatsHandler(&statsHandler{}),
+		grpc.StreamInterceptor((&activeStreamCounter{}).Intercept),
+	)
+
+	// TODO(streaming): add gRPC services to srv here
+
+	return &Handler{
+		srv:      srv,
+		listener: &chanListener{addr: addr, conns: make(chan net.Conn)},
+	}
+}
+
+// Handler implements a handler for the rpc server listener, and the
+// agent.Component interface for managing the lifecycle of the grpc.Server.
+type Handler struct {
+	srv      *grpc.Server
+	listener *chanListener
+}
+
+// Handle the connection by sending it to a channel for the grpc.Server to receive.
+func (h *Handler) Handle(conn net.Conn) {
+	h.listener.conns <- conn
+}
+
+func (h *Handler) Run() error {
+	return h.srv.Serve(h.listener)
+}
+
+func (h *Handler) Shutdown() error {
+	h.srv.Stop()
+	return nil
+}
+
+// chanListener implements net.Listener for grpc.Server.
+type chanListener struct {
+	conns chan net.Conn
+	addr  net.Addr
+}
+
+// Accept blocks until a connection is received from Handle, and then returns the
+// connection. Accept implements part of the net.Listener interface for grpc.Server.
+func (l *chanListener) Accept() (net.Conn, error) {
+	return <-l.conns, nil
+}
+
+func (l *chanListener) Addr() net.Addr {
+	return l.addr
+}
+
+// Close does nothing. The connections are managed by the caller.
+func (l *chanListener) Close() error {
+	return nil
```

Comment:
I had the same question when I saw it here: https://github.com/hashicorp/consul/blob/streaming-rpc-final/agent/consul/rpc.go#L59-L78

I assumed it was because yamux manages the connections, so closing it would be incorrect, as other requests may be using this connection. Is that not the case?

Reply:
> t I think in tests at least this would end up leaking the goroutine in agent.go above that runs the Run method since Accept will never return again after this closes.

Won't the call to `Handler.Shutdown` , which calls `grpc.Server.Shutdown` cause the Accept to exit and the goroutine to exit? I haven't looked at the code, but that was my understanding.

Reply:
I ended up running into this problem while writing tests for #8667, that PR contains a fix.